File: solutions/4.median-of-two-sorted-arrays.py
#
# @lc app=leetcode id=4 lang=python3
#
# [4] Median of Two Sorted Arrays
#

# @lc code=start
import sys
from typing import List


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        small = nums1 if len(nums1) <= len(nums2) else nums2
        big = nums2 if len(nums1) <= len(nums2) else nums1

        total = len(small) + len(big)
        half = (total + 1) // 2 # round up in case of the odd total length

        result = 0.0
        left = 0
        right = len(small)
        while left <= right:
            small_middle = (left + right) // 2
            big_middle = half - small_middle

            small_left = small[small_middle - 1] if small_middle > 0 else -sys.maxsize
            small_right = small[small_middle] if small_middle < len(small) else sys.maxsize
            big_left = big[big_middle - 1] if big_middle > 0 else -sys.maxsize
            big_right = big[big_middle] if big_middle < len(big) else sys.maxsize

            if small_left <= big_right and big_left <= small_right:
                if total % 2:
                    result = max(small_left, big_left)
                else:
                    result = (max(small_left, big_left) + min(small_right, big_right)) / 2.0
                break
            elif small_left > big_right:
                right = small_middle - 1
            else:
                left = small_middle + 1

        return result

# @lc code=end

File: solutions/4.median-of-two-sorted-arrays.java
/*
 * @lc app=leetcode id=4 lang=java
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int[] small = nums1.length <= nums2.length ? nums1 : nums2;
        int[] big = nums1.length <= nums2.length ? nums2 : nums1;

        double result = 0;
        int total = small.length + big.length;
        int half = (total + 1) / 2;
        int left = 0;
        int right = small.length;
        while (left <= right) {
            int smallMiddle = (left + right) >>> 1;
            int bigMiddle = half - smallMiddle;

            int smallLeft = smallMiddle > 0 ? small[smallMiddle - 1] : Integer.MIN_VALUE;
            int smallRight = smallMiddle < small.length ? small[smallMiddle] : Integer.MAX_VALUE;
            int bigLeft = bigMiddle > 0 ? big[bigMiddle - 1] : Integer.MIN_VALUE;
            int bigRight = bigMiddle < big.length ? big[bigMiddle] : Integer.MAX_VALUE;

            if (smallLeft <= bigRight && bigLeft <= smallRight) {
                if (total % 2 != 0) {
                    result = Math.max(smallLeft, bigLeft);
                } else {
                    result = (Math.max(smallLeft, bigLeft) + Math.min(smallRight, bigRight)) / 2.0;
                }
                break;
            } else if (smallLeft > bigRight) {
                right = smallMiddle - 1;
            } else {
                left = smallMiddle + 1;
            }
        }

        return result;
    }
}
// @lc code=end

